src/h3/mod.rs

```diff
@@ -1165,6 +1146,124 @@ impl Connection {
     /// [`recv_dgram()`]: struct.Connection.html#method.recv_dgram
     /// [`close()`]: ../struct.Connection.html#method.close
     pub fn poll(&mut self, conn: &mut super::Connection) -> Result<(u64, Event)> {
+        let mut event = None;
+
+        // When no primary event is issued (e.g. due to stream or DATAGRAM poll
+        // thresholds), we need to fallback to the first stream or datagram
+        // event regardless of thresholds.
+        //
+        // For example, if the DATAGRAM threshold is reached, but no stream
+        // events are being generated (e.g. because there are no readable stream),
+        // we should return a DATAGRAM event if available.
+        let mut fallback_event = None;
+
+        // Both stream and DATAGRAM event limits have been reached,
+        // so reset and start again.
+        if self.dgram_event_count == self.dgram_poll_threshold &&
+            self.stream_event_count == self.stream_poll_threshold
+        {
+            self.dgram_event_count = 0;
+            self.stream_event_count = 0;
+        }
+
+        let do_poll = |h3: &mut Connection, _: &mut crate::Connection, s, ev| {
+            match ev {
+                Event::Headers { .. } | Event::Data => {
+                    if h3.stream_event_count < h3.stream_poll_threshold {
+                        event = Some((s, ev));
+
+                        h3.stream_event_count += 1;
+                    } else if fallback_event.is_none() {
+                        fallback_event = Some((s, ev));
+                    }
+                },
+
+                Event::Datagram => {
+                    if h3.dgram_event_count < h3.dgram_poll_threshold {
+                        event = Some((s, ev));
+
+                        h3.dgram_event_count += 1;
+                    } else if fallback_event.is_none() {
+                        fallback_event = Some((s, ev));
+                    }
+                },
+
+                _ => event = Some((s, ev)),
+            }
+
+            if event.is_some() {
```

Comment:
is this supposed to be is_none()?

Reply:
This is on purpose. The idea being that we can return early as soon as we got an event.